File: 2904-shortest-and-lexicographically-smallest-beautiful-string/2904-shortest-and-lexicographically-smallest-beautiful-string.py
class Solution:
    def shortestBeautifulSubstring(self, s: str, k: int) -> str:
        n = len(s)

        ones = 0
        ans = ""
        l  = 0

        for r in range(n):
            if s[r]=='1' : ones+=1

            while l <= r and (ones > k or (ones== k and s[l]=='0')):
                if s[l]=='1' : ones -= 1
                l+=1

            if ones == k:
                sub = s[l : r+1]
                if (not ans or 
                len(sub)<len(ans) or 
                (len(sub)==len(ans) and sub<ans)):
                    ans = sub
        
        return ans




       

        

File: 2904-shortest-and-lexicographically-smallest-beautiful-string/2904-shortest-and-lexicographically-smallest-beautiful-string.cpp
class Solution {
public:
    string shortestBeautifulSubstring(string s, int k) {
        int n = s.size();

        int l = 0 , ones = 0;
        string res = "";

        for (int r = 0 ; r < n ; r++){
            if (s[r] == '1') ones++;

            while (l<r && (ones > k || s[l]=='0')){
                if (s[l]=='1') ones--;
                l++;
            }

            if (ones == k){
                string sub = s.substr(l , r-l+1);

                if (res=="" || sub.size()<res.size() || (sub.size() == res.size() && sub < res)){
                    res = sub;
                }
            }
        }
        return res;

    }
};